R/read_target_file.R

```diff
@@ -0,0 +1,116 @@
+#' - `chr`: read in (`csv`) or coerce (`parquet`) all columns to character.
+#' - `none`: No coercion. Use `arrow` `read_*` function defaults.
+#'
+#' @returns  a tibble of contents of the target data file.
+#' @export
+#'
+#' @examples
+#' # download example hub
+#' hub_path <- withr::local_tempdir()
+#' example_hub <- "https://github.com/hubverse-org/example-complex-forecast-hub.git"
+#' gert::git_clone(url = example_hub, path = hub_path)
+#' # read in time-series file
+#' read_target_file("time-series.csv", hub_path)
+#' read_target_file("time-series.csv", hub_path, coerce_types = "chr")
+#' # read in oracle-output file
+#' read_target_file("oracle-output.csv", hub_path)
+#' read_target_file("oracle-output.csv", hub_path, coerce_types = "chr")
+read_target_file <- function(target_file_path, hub_path,
+                             coerce_types = c("hub", "chr", "none"),
+                             date_col = NULL,
+                             na = c("NA", "")) {
+  checkmate::assert_character(target_file_path, len = 1)
+  checkmate::assert_character(hub_path, len = 1)
+
+  coerce_types <- rlang::arg_match(coerce_types)
+  target_type <- guess_target_type_from_path(target_file_path)
+
+  full_path <- abs_file_path(target_file_path, hub_path, subdir = "target-data")
+
+  if (!fs::file_exists(full_path)) {
+    rel_path <- rel_file_path(target_file_path, hub_path, subdir = "target-data") # nolint: object_usage_linter
+    cli::cli_abort("No file exists at path {.path {rel_path}}")
+  }
+
+  file_ext <- fs::path_ext(target_file_path)
+  valid_ext <- c("csv", "parquet")
+  if (!file_ext %in% valid_ext) {
+    cli::cli_abort("File cannot be read. File extension must be one of {.val {valid_ext}}, not {.val {file_ext}}")
+  }
+
+  # Determine schema if coercion is required
+  schema <- NULL
+  if (coerce_types == "hub") {
+    schema <- switch(target_type,
+      `time-series` = hubData::create_timeseries_schema(
+        hub_path = hub_path, date_col = date_col, na = na
+      ),
+      `oracle-output` = hubData::create_oracle_output_schema(
+        hub_path = hub_path, na = na
+      )
+    )
+  }
+
+  # Read file
+  tbl <- switch(file_ext,
+    csv = {
+      out <- arrow::read_csv_arrow(full_path, col_types = schema)
+      if (coerce_types == "chr") {
+        out <- hubData::coerce_to_character(out)
+      }
+      out
+    },
+    parquet = {
+      if (coerce_types == "hub") {
+        arrow::read_parquet(full_path, schema = schema)
+      } else if (coerce_types == "chr") {
+        arrow::read_parquet(full_path) |>
+          hubData::coerce_to_character()
+      } else {
+        arrow::read_parquet(full_path)
+      }
+    }
+  )
+
+  # Convert to tibble
+  tbl <- tibble::as_tibble(tbl)
+
+  # Extract partition values and bind them
+  partition_row <- extract_partition_df(
+    target_file_path, schema, strict = TRUE
+  )
+  if (!is.null(partition_row)) {
+    tbl <- dplyr::bind_cols(tbl, partition_row)
```

Comment:
This confused me a bit because I was trying to figure out why we were binding cols with a row of data.

I now understand that it's because you will only ever get one row of data with this function, but I would prefer to name it as "partition_df" to avoid that confusion. 

```suggestion
  partition_df <- extract_partition_df(
    target_file_path, schema, strict = TRUE
  )
  if (!is.null(partition_df)) {
    tbl <- dplyr::bind_cols(tbl, partition_df)
```